xwiki-platform-core/xwiki-platform-javascript/xwiki-platform-javascript-importmap/src/main/java/org/xwiki/javascript/importmap/internal/JavascriptImportmapResolver.java

```diff
@@ -0,0 +1,187 @@
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.javascript.importmap.internal;
+
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import jakarta.inject.Inject;
+import jakarta.inject.Singleton;
+
+import org.slf4j.Logger;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.extension.Extension;
+import org.xwiki.extension.repository.CoreExtensionRepository;
+import org.xwiki.extension.repository.InstalledExtensionRepository;
+import org.xwiki.rendering.block.Block;
+import org.xwiki.rendering.block.RawBlock;
+import org.xwiki.webjars.WebJarsUrlFactory;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.json.JsonMapper;
+
+import static com.fasterxml.jackson.databind.MapperFeature.SORT_PROPERTIES_ALPHABETICALLY;
+import static org.apache.commons.lang3.exception.ExceptionUtils.getRootCauseMessage;
+import static org.xwiki.rendering.syntax.Syntax.HTML_5_0;
+
+/**
+ * Resolve the importmap from the webjar declarations, using the {@code "xwiki.javascript.modules.importmap"} pom
+ * property.
+ *
+ * @version $Id$
+ * @since 17.10.0RC1
+ */
+@Component(roles = JavascriptImportmapResolver.class)
+@Singleton
+public class JavascriptImportmapResolver
+{
+    private static final ObjectMapper OBJECT_MAPPER = JsonMapper.builder()
+        // Keeps the output deterministic.
+        .configure(SORT_PROPERTIES_ALPHABETICALLY, true)
+        .build();
+
+    private static final String JAVASCRIPT_IMPORTMAP_PROPERTY = "xwiki.javascript.modules.importmap";
+
+    /**
+     * Cache lock to prevent race conditions when retrieving cached values.
+     */
+    private final Object cacheLock = new Object();
+
+    /**
+     * The cached value, concurrent operation on this field must use the {@link #cacheLock} object for synchronization.
+     */
+    private Block cachedValue;
+
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private InstalledExtensionRepository installedExtensionRepository;
+
+    @Inject
+    private CoreExtensionRepository coreExtensionRepository;
+
+    @Inject
+    private WebJarsUrlFactory webJarsUrlFactory;
+
+    /**
+     * @return the block to render from the importmap, possibly cached
+     */
+    public Block getBlock()
+    {
+        synchronized (this.cacheLock) {
+            if (this.cachedValue == null) {
+                compute();
+            }
+            return this.cachedValue;
+        }
+    }
+
+    /**
+     * Clear the cache of the resolver.
+     */
+    public void clearCache()
+    {
+        synchronized (this.cacheLock) {
+            this.cachedValue = null;
+        }
+    }
+
+    private void compute()
```

Comment:
Might make more sense to have this per wiki (some extension might be installed in a specific wiki, and it would not make much sense to have it in the import map of another wiki).

Reply:
Good point.
I'm expecting this component to be instantiated once per wiki since the extension is not installed at the farm level.
So I guess what's missing is a good way to load only the extensions installed on the current wiki.
Not sure what's the best way to achieve this, though.

Reply:
>  Not sure what's the best way to achieve this, though.

Checking the others #getInstalledExtensions methods.

Reply:
Ah right, the `namespace` parameter can be filled with the wiki id?